src/app/services/transmart-services/transmart-external-job-resource.service.ts

```diff
@@ -0,0 +1,260 @@
+    this._customExportJobName = value;
+  }
+
+  get endpointUrl(): string {
+    return this._endpointUrl;
+  }
+
+  set endpointUrl(value: string) {
+    this._endpointUrl = value;
+  }
+
+  get autosaveSubjectSets(): boolean {
+    return this._autosaveSubjectSets;
+  }
+
+  set autosaveSubjectSets(value: boolean) {
+    this._autosaveSubjectSets = value;
+  }
+
+  // ------------------------------------- export job calls -------------------------------------
+
+  /**
+   * Make a post http request
+   * @param urlPart - the part used in baseUrl/urlPart
+   * @param body
+   * @param responseField
+   * @returns {Observable<any | any>}
+   */
+  private postCall(urlPart, body, responseField) {
+    const url = `${this.endpointUrl}/${urlPart}`;
+    if (responseField) {
+      return this.http.post(url, body).pipe(
+        map(res => res[responseField]),
+        catchError(error => {
+          ErrorHelper.handleError(error);
+          return observableOf(error);
+        })
+      );
+    } else {
+      return this.http.post(url, body).pipe(
+        catchError(error => {
+          ErrorHelper.handleError(error);
+          return observableOf(error);
+        })
+      );
+    }
+  }
+
+  /**
+   * Make a get http request
+   * @param urlPart - the part used in baseUrl/urlPart
+   * @param responseField
+   * @returns {Observable<any | any>}
+   */
+  private getCall(urlPart, responseField) {
+    const url = `${this.endpointUrl}/${urlPart}`;
+    if (responseField) {
+      return this.http.get(url).pipe(
+        map((res) => res[responseField]),
+        catchError((error: HttpErrorResponse) => {
+          ErrorHelper.handleError(error);
+          return observableOf(error);
+        })
+      );
+    } else {
+      return this.http.get(url).pipe(
+        catchError((error: HttpErrorResponse) => {
+          ErrorHelper.handleError(error);
+          return observableOf(error);
+        })
+      );
+    }
+  }
+
+  /**
+   * Get the current user's existing export jobs
+   * @returns {Observable<any[]>}
+   */
+  getAllJobs(): Observable<TransmartExternalJob[]> {
+    const urlPart = 'jobs';
+    const responseField = 'jobs';
+    return this.getCall(urlPart, responseField);
+  }
+
+  /**
+   * Get a status of the job
+   * @param jobId
+   * @returns {Observable<ExportJob>}
+   */
+  getJobStatus(jobId: string): Observable<TransmartExternalJob> {
+    const urlPart = `jobs/status/${jobId}`;
+    const responseField = 'exportJob';
+    return this.getCall(urlPart, responseField);
+  }
+
+  /**
+   * Create and run an export job for the current user, with a given name
+   *
+   * @param {string} jobName
+   * @param {Constraint} constraint
+   * @returns {Observable<ExportJob>}
+   */
+  runJob(jobName: string, constraint: Constraint): Observable<TransmartExternalJob> {
+    const urlPart = `jobs/create`;
+    const responseField = 'job';
+    let targetConstraint = constraint;
+    if (this.autosaveSubjectSets &&
```

Comment:
I don't really understand what happens here, but whatever it is, it does not seem like something that should be the responsibility of this service.

Reply:
First step would be to create a separate function that transforms the constraint in one where the subject level part is replaced with the subject set constraint. And then decide which service is responsible... the transmart resource service, or, perhaps a separate service: maybe a superclass for both?